### DISCUSSION TITLE
Release `rust-bitcoin` v0.32.0

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#2375] [General]
Shall we shoot for end of March?

Please use milestone 0.32.0 to add PRs and issues that you'd like to see get into the next release.

**Comment by tcharding** [comment]
This is just based on my current view of the world, those with write perms please update with your own views. Anyone else who wants to see something in the next release please post and I"ll update the message above.

**Comment by tcharding** [comment]
~Added a project: https://github.com/orgs/rust-bitcoin/projects/8~ (Deleted)

**Comment by tcharding** [comment]
If there are any issues you guys would like to see done before next release can you add them to the project backlog please and I'll pick them up.

**Reply by Kixunil** [reply]
I guess let's just use milestone.

**Reply by tcharding** [reply]
Oh, is milestone better than using a project? So far, I have not managed to coerce the github project feature to give me anything of value. But perhaps I just don't know how to project manage. I can switch everything to milestones if that is better?

**Reply by tcharding** [reply]
What I wanted was a way to visually link dependencies between issues/PRs i.e. this blocks that. But I couldn't work this out.

**Reply by Kixunil** [reply]
Milestone is easier to work with than editing a discussion and you can filter by it.

**Reply by tcharding** [reply]
Oh I mean the "projects" feature of github: https://github.com/orgs/rust-bitcoin/projects/8

I did this issue with the links to try and help visualize whats got to be done instead of having a list. I'm 100% open to better ways to do this, I'm not doing the project management because its fun, just because I think it needs doing. Best/easiest/cleanest/fastest way possible is good for me.

**Reply by Kixunil** [reply]
I really just need a list of issues/PRs that need to be done. And setting a milestone is faster than copy-pasting.

**Reply by tcharding** [reply]
Added the milestone, I'm going to attempt to use projects for the other crate releases that are required for `rust-bitcoin` v0.32.0 and see if this helps track PR/issue dependencies.

**Comment by tcharding** [comment]
I am becoming increasingly concerned that we are going to blow wildly past this end-of-march release date. We are doing invasive changes to hex stuff (mainly error code) that effects `units`, and `units` must be released before we can release `rust-bitcoin v0.32.0` - with the slowness of our dev-review-merge cycle I believe we need to start releasing the lower crates well before the release date of `rust-bitcoin` - if we assume a two week RC window for `rust-bitcoin` then this further shortens the timeframe. IMO we need to release `units` within the next two weeks.

**Reply by Kixunil** [reply]
Honestly I'd rather miss the deadline. What's the point of releasing broken stuff? We didn't promise anyone anything and nobody is paying us to release on specific date.

**Reply by apoelstra** [reply]
>What's the point of releasing broken stuff?

* To get feedback from people on things we may not realize are broken.
* To have some tangible sense of progress.
* To keep us disciplined regarding scope of releases.

**Reply by Kixunil** [reply]
> To get feedback from people on things we may not realize are broken.

Of course, I'm talking about knowingly broken things.

Debian has "when ready, roughly every two year" policy and it seems to work well from what I can tell. I don't see why we can't have something similar.

**Reply by apoelstra** [reply]
I don't mind being "roughly" every six months, but if we routinely slip by multiple months, I think we need to be more agressive about "just shipping" stuff.

It's true that we have some knowingly-broken things, but most of these knowingly-broken things are ancient parts of the codebase and it'll take multiple release cycles to fix them all.

Personally I would rather have multiple "progressively-less-broken" releases than widely-interspersed releases that do a ton at once and take a ton of effort and ugliness to update. (Though @stevenroose has argued to me elsewhere that frequent releases can also be a PITA since multiple other library authors feel obligated to update their crates immediately, even if in-principle we expect end-users to be batching updates. But I think part of why this is painful is that we're breaking too much at once.)

**Reply by Kixunil** [reply]
Or maybe we need longer freeze periods, better planning etc.

**Reply by tcharding** [reply]
With how unknown the work involved in crate smashing is (who know hex would take the amount of time it is taking) I'm not sure we can plan much better, at least I don't know how, without reading a book on project management (which I don't plan on doing).

But it seems if we get a push that "we realy want XYZ in next release" then as the release approaches we need to soften up on "XYZ must be perfect". I think we should follow more strictly the old mantra "release early, release often" instead of trying to make everything perfect.

**Reply by Kixunil** [reply]
Maybe we should define our priorities. My idea was to try releasing `units` as if it was 1.0, then wait for a year if there's a strong motivation for API break, if yes, break and restart the counter, if not, release 1.0. And obviously if it's a year we want to do it ASAP so that we don't stabilize super late. Now the other approach would be to release now and then release `units` 0.2 soon after when we do what we want. But if we do that it won't get much testing unless we also release `bitcoin` 0.33. This just adds a lot of churn and maybe people will be unwilling to upgrade again.

So the question is are you even on board with this idea? What are the alternatives?

**Reply by apoelstra** [reply]
>My idea was to try releasing units as if it was 1.0,

If we do this then it will take forever to release `units` and it will be way harder to move forward on everything else. We should just release `units` 0.1 and start using it. Nobody will depend directly on it so we can do as many releases as we want and it won't matter until the next release of rust-bitcoin, and churn after that also won't matter until the *next* release of rust-bitcoin.

Once we are close to `units` 1.0 (same with most all of the other new crates) we can start encouraging people to directly use these crates and then we need to be careful about them.

**Comment by tcharding** [comment]
I think we should stick to the six months release cadence we have achieved recently, to do so we have to release all the other crates in the `rust-bitcoin` repo or we have to stop patching the manifest like we do. As I already said I think we should release early, release often.

**Comment by tcharding** [comment]
Its two weeks until the end of March crew, we have a whole bunch of open PRs and issues that are marked milestone 0.32 and we need to get cracking on releasing the stack of crates. Could anyone with review time please sort by milestone and throw some ACKs.

@sanket1729 I have only marked the hex stuff as priority because otherwise I'd mark all the milestone 0.32 stuff but that seems redundant.

**Comment by tcharding** [comment]
I've triaged all open PRs of my own and non-maintainers and added the "milestone 0.32.0", converted to draft (or asked author to do so).

Shall we soft feature freeze soon and try and get this release done?

**Reply by apoelstra** [reply]
Let's do it.

I think we can do a "soft" feature-freeze now. Let's get existing PRs in and anything urgent that comes up and anything new should be deferred until the next release.

**Reply by tcharding** [reply]
"existing PRs" is too broad IMO - there are open non-draft things that are either far off being done or require a non-trivial amount of work/review. That's why I attempted to encourage contributors to draft such PRs. But I intentionally did not comment on maintainer PRs because
- @Kixunil is well aware of what is going on
- I didn't want to upstet @stevenroose and couldn't think of a way to comment that didn't risk doing so :)

**Reply by apoelstra** [reply]
I think @stevenroose is going to be upset no matter what about us cutting releases too often. I talked to him in person briefly at Advancing Bitcoin and suggested we

* Release an rc first (which we do)
* Update miniscript and bdk to use the rc so we can feel out the pain points (which I think we do, but usually just one of us does it and figures "if it's possible at all to do the update then it's fine")
* Fix anything that's really annoying or feels gratuitiously changed.
* Write a migration guide.

Steven also hasn't updated rust-bitcoin since 0.30 which was the absolute worst. It didn't support hex, required renaming Script to ScriptBuf, broke addresses in a hard-to-discover way, etc etc etc. So bear in mind that this is his reference point for complaining.

**Reply by apoelstra** [reply]
But I wouldn't bother tiptoeing around his PRs in the hopes he doesn't notice the release :P

**Reply by tcharding** [reply]
> Steven also hasn't updated rust-bitcoin since 0.30 which was the absolute worst.

I agree this is the main problem.

> Fix anything that's really annoying or feels gratuitously changed.

Sounds good but feels to me like during PR/review, or at worst raising an issue soon after merge, is the time to do this not during the RC cycle - but I'm open to it.

**Reply by apoelstra** [reply]
>Sounds good but feels to me like during PR/review, or at worst raising an issue soon after merge, is the time to do this not during the RC cycle - but I'm open to it.

If you are willing to try updating bdk after every PR then be my guest! Although this may give an inaccurately rosy picture of how hard the final update will be, since you'll be doing it in small chunks where you only have to fix a single thing at a time (and a thing that happened in a PR you just reviewed).

**Reply by tcharding** [reply]
Oh I meant the gratuitous changes bit - which by definition wouldn't be considered gratuitous if I'd just done the PR :) Jokes aside I can most definitely attempt to do the bdk update during the RC cycle of `rust-bitcoin` - I'm pretty sure I did this last release. FTR since before last release I have had the personal opinion that the day we drop a `rust-bitcoin` all the crates in our github org should be ready to update also - that means, to me personally, a draft PR is open already.

**Reply by apoelstra** [reply]
Agreed, let's say that it's a prereq that we have a draft PR open for rust-miniscript, bdk, bech32 and rust-elements.

I'm not sure exactly how to do bdk because it depends on rust-miniscript -- maybe we should do an actual rc of rust-miniscript which uses the rc of rust-bitcoin?

I'd like to also have a draft PR open for BitVM but they're currently several versions behind (and off on a temporary fork of Steven's with some stuff we can't upstream during feature freeze). So for 0.33 let's target that.

**Reply by tcharding** [reply]
> Agreed, let's say that it's a prereq that we have a draft PR open for rust-miniscript, bdk, bech32 and rust-elements.

I won't remember where this thread is later on to reference it but can we agree that I do miniscript and bdk and you do elements? bech32 is a mistake in your comment I believe, did you mean something else? 

> I'm not sure exactly how to do bdk because it depends on rust-miniscript 

I can do the PRs to include patch sections in the manifest to point to the correct branches then you can pull it down locally to check them (if you want), CI clearly won't work though (because of `cargo --locked`).

> I'd like to also have a draft PR open for BitVM 

I've checked in with them: https://github.com/BitVM/BitVM/issues/23

**Reply by apoelstra** [reply]
Yeah, sorry, bech32 is a mistake.

But yes, I will do elements (and maybe elements-miniscript and rust-simplicity, if I get practiced enough at using `[patch]` sections).

**Reply by tcharding** [reply]
Pro tip which I often forget, you have to have patch sections for all the crates in `rust-bitcoin` not just one for `bitcoin` (ie, `units`, `internals` etc).

**Comment by tcharding** [comment]
Lets go!

- https://github.com/rust-bitcoin/rust-bitcoin/pull/2596
- https://github.com/rust-bitcoin/rust-bitcoin/pull/2595

**Comment by tcharding** [comment]
The first release candidate should have dropped already and we have [a whole bunch of open PRs](https://github.com/rust-bitcoin/rust-bitcoin/pulls?q=is%3Aopen+is%3Apr+milestone%3A0.32.0) still. Any maintainers available to help out reviewing would be super appreciated (@apoelstra will totally buy you a beer when he sees you :)

**Comment by tcharding** [comment]
## TODO list for `rust-bitcoin v0.32.0-rc1`

- [x] Merge and release `secp256k1-sys` (https://github.com/rust-bitcoin/rust-secp256k1/pull/688)
- [x]  Merge and release `rust-secp256k1` (https://github.com/rust-bitcoin/rust-secp256k1/pull/685)
- [x] Merge #2636 
- [x] Release `internals` 
- [x] Release `base58ck`  - (ready soon as `internals` is released)
- [x] Merge #2514
- [x] Merge #2541 
- [x] Merge #2651 and release `units` 
- [x] Follow up to 2514: #2650
- [x] Merge #2652
- [x] https://github.com/rust-bitcoin/rust-bitcoin/pull/2655
- [x] Merge #2657 and release `bitcoin-0.32.0-rc1`

**Reply by tcharding** [reply]
cc @apoelstra just to spam you because I can't remember where I mentioned base58ck release, `cargo publish --dry-run` works for me locally - tag and publish `base58ck` please.

**Reply by tcharding** [reply]
Three tasks to go boys. Lets get the RC out this week, come on!

**Comment by tcharding** [comment]
Hey crew,

I've upgraded all crates up to and including `bdk` with the `rc1` release. The following PRs fell out of it:
- https://github.com/rust-bitcoin/rust-bitcoin/pull/2676
- https://github.com/rust-bitcoin/rust-bitcoin/pull/2674
- https://github.com/rust-bitcoin/rust-bitcoin/pull/2666

All the test branches are pushed up to my forks and a draft PR is opened on each repo. Branch name is `test-bitcoin` everywhere. Here are a couple of the more important upgrades:

- https://github.com/rust-bitcoin/rust-miniscript/pull/662
- https://github.com/bitcoindevkit/bdk/pull/1400

For a more complete list see the top entries of https://github.com/pulls?q=is%3Aopen+is%3Apr+author%3Atcharding+archived%3Afalse+Depend+on+bitcoin
